xds/src/test/java/io/grpc/xds/FallbackLbTest.java

```diff
@@ -0,0 +1,377 @@
+        .build());
+
+    assertThat(helpers1).hasSize(1);
+    assertThat(balancers1).hasSize(1);
+    Helper helper1 = helpers1.get(0);
+    LoadBalancer balancer1 = balancers1.get(0);
+    verify(balancer1).handleResolvedAddresses(ResolvedAddresses.newBuilder()
+        .setAddresses(eags11)
+        .setAttributes(Attributes.newBuilder()
+            .set(ATTR_LOAD_BALANCING_CONFIG, ImmutableMap.of("fallback_1_option", "yes")).build())
+        .build());
+
+    SubchannelPicker picker1 = mock(SubchannelPicker.class);
+    helper1.updateBalancingState(CONNECTING, picker1);
+    verify(helper).updateBalancingState(CONNECTING, picker1);
+
+    EquivalentAddressGroup eag121 = new EquivalentAddressGroup(mock(SocketAddress.class));
+    List<EquivalentAddressGroup> eags12 = ImmutableList.of(eag121);
+    String lbConfigRaw12 = "{"
+        + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
+        + "\"fallbackPolicy\" : [{\"fallback_1\" : { \"fallback_1_option\" : \"no\"}}]"
+        + "}";
+    @SuppressWarnings("unchecked")
+    Map<String, ?> lbConfig12 = (Map<String, ?>) JsonParser.parse(lbConfigRaw12);
+    fallbackLb.handleResolvedAddresses(ResolvedAddresses.newBuilder()
+        .setAddresses(eags12)
+        .setAttributes(Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig12).build())
+        .build());
+
+    verify(balancer1).handleResolvedAddresses(ResolvedAddresses.newBuilder()
+        .setAddresses(eags12)
+        .setAttributes(Attributes.newBuilder()
+            .set(ATTR_LOAD_BALANCING_CONFIG, ImmutableMap.of("fallback_1_option", "no")).build())
+        .build());
+
+    verify(balancer1, never()).shutdown();
+    assertThat(helpers2).isEmpty();
+    assertThat(balancers2).isEmpty();
+
+    // change fallback policy to fallback_2
+    EquivalentAddressGroup eag211 = new EquivalentAddressGroup(mock(SocketAddress.class));
+    EquivalentAddressGroup eag212 = new EquivalentAddressGroup(mock(SocketAddress.class));
+    List<EquivalentAddressGroup> eags21 = ImmutableList.of(eag211, eag212);
+    String lbConfigRaw21 = "{"
+        + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
+        + "\"fallbackPolicy\" : [{\"fallback_2\" : { \"fallback_2_option\" : \"yes\"}}]"
+        + "}";
+    @SuppressWarnings("unchecked")
+    Map<String, ?> lbConfig21 = (Map<String, ?>) JsonParser.parse(lbConfigRaw21);
+    fallbackLb.handleResolvedAddresses(ResolvedAddresses.newBuilder()
+        .setAddresses(eags21)
+        .setAttributes(Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig21).build())
+        .build());
+
+    verify(balancer1).shutdown();
+    assertThat(helpers2).hasSize(1);
+    assertThat(balancers2).hasSize(1);
+    Helper helper2 = helpers2.get(0);
+    LoadBalancer balancer2 = balancers2.get(0);
+    verify(balancer1, never()).handleResolvedAddresses(ResolvedAddresses.newBuilder()
```

Comment:
The behavior here should be the same regardless there's `GracefulSwitchLoadBalancer' in place of hard switch in place.

Reply:
Then you don't need to verify anything more than `verify(balancer1).shutdown()` for `balancer1`. Anything more than that is excessive and is a reproduce of implementation. Tests do not aim to cover everything in implementation, but only observable checkpoints. But sometimes this can be vague, I am fine with it..